### ISSUE TITLE
Binary search is used with incorrectly-sorted array

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/regex#75]
The “dynamic” matching of `CharClass` [uses a binary search](https://github.com/rust-lang/regex/blob/399758aeae3dcab382b0af5fa9964c1e32066dda/src/vm.rs#L236) within `char` ranges, which relies on the input being sorted. The input is indeed sorted in its “natural” order, but the comparison function in case-insensitive mode uses a different order.

This leads to incorrect results:

```rust
assert!(Regex::new(r"(?i)[a_]+$").unwrap().is_match("A_"));
```

The above fails, because `_` in ASCII is between upper-case and lower-case letters. The comparison function maps the `'a'..'a'` range to its upper case `'A'..'A'`, which has a different order relative to `'_'..'_'`.

Compare with e.g. the code below, which succeeds.
```rust
assert!(Regex::new(r"(?i)[a=]+$").unwrap().is_match("A="));
```

The comparison function has a `FIXME` to move the case mapping outside of it and have the `Vec` of ranges be already mapped. I assume this was intended for performance, but I think it’ll also fix this issue. I’ll try it.

**Comment by BurntSushi** [comment]
Nice find! If you find an easy fix, that would be awesome. Otherwise, I am in the process of rewriting the parser with the intention of fixing a bunch of little bugaboos like this. (Although I didn't realize this caused buggy behavior!)

**Comment by SimonSapin** [comment]
I have a fix, but I’m also tracking down another bug. Let we write up an issue.

**Comment by SimonSapin** [comment]
#75 is that issue. #78 has fixes for both and #55.

**Comment by BurntSushi** [comment]
Awesome! I'm currently out, but I'll review as soon as I get home and get this merged. Thank you!
